### PR TITLE
add fe proxy

### PR DESCRIPTION
**Pull request by yandongxiao**
Signed-off-by: yandongxiao <yandongxiao@starrocks.com>

### CODE REVIEW

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml [hunk 1]
```diff
@@ -408,3 +408,54 @@ configMaps: []
...
+    tag: 1.24.0
+  resources:
+    limits:
+      cpu: 1
+      memory: 2Gi
+    requests:
+      cpu: 1
+      memory: 2Gi
+  # set the resolver for nginx server
+  resolver: kube-dns.kube-system.svc.cluster.local
```
Comment (kevincai) [inline helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:428]: can we make this default?

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml [hunk 2]
```diff
@@ -408,3 +408,54 @@ configMaps: []
   #     key: |
   #       this is the content of the configmap
   #       when mounted, key will be the name of the file
+
+# specify the fe proxy deploy or not.
+feProxy:
+  # specify the fe proxy deploy or not.
+  enabled: true
+  replicas: 1
+  image:
```
Comment (kevincai) [inline helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:417]: can we make this image section omit by default?

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml [hunk 3]
```diff
@@ -408,3 +408,54 @@ configMaps: []
...
+  tolerations: []
+    # - key: "key"
+    #   operator: "Equal|Exists"
+    #   value: "value"
+    #   effect: "NoSchedule|PreferNoSchedule|NoExecute(1.6 only)"
+  # This ConfigMap store the nginx.conf for fe proxy.
+  # If configMapName is empty, operator will use its default nginx.conf.
+  # You can create a ConfigMap with the key nginx.conf to override the default nginx.conf, and please
+  # make sure nginx server is listening on port 8080
+  configMapInfo:
```
Comment (kevincai) [inline helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:460]: any specific configmap we needed to make it running for now?

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml (hunk 1 above)
Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:428]: the default value is also kube-dns.kube-system.svc.cluster.local. I will set it to empty string in values.yaml

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml (hunk 2 above)
Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:417]: ok, I will check these fields, and make them have default values.

File: helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml (hunk 3 above)
Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:460]: operator will create a confimap to store the nginx.conf, mount it to pod

Comment (kevincai) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:460]: we can even generate.the conf on the fly in entrypoint with necessary env var injected by operator.

Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:462]: > if we are not going to allow user to provide their own nginx conf, is this configmap needed in crd?

no, shall we delete it from values.yaml ?

Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:462]: operator allow users to use their own nginx.conf, in case our implementation is not comlete.

Comment (kevincai) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:462]: I don't think so, I can't think of any cases that the user needs to modify this nginx.conf. In order not to build our own nginx img, we can do the nginx conf mounted from configmap by the operator. the img is necessary in case they can't access the img from public network.

Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml:462]: ok, I will remove it.

File: helm-charts/charts/kube-starrocks/values.yaml [hunk 4]
```diff
@@ -66,6 +66,7 @@ starrocks:
   initPassword:
     enabled: false
     password: ""
+    # The secret name that contains password, the key of the secret is "password", and you should create it first.
```
Comment (kevincai) [inline helm-charts/charts/kube-starrocks/values.yaml:69]: should move this into a separate PR.

File: deploy/operator.yaml [hunk 5]
```diff
@@ -23,168 +23,52 @@ rules:
...
   - statefulsets
   verbs:
-  - create
-  - delete
-  - get
-  - list
-  - patch
-  - update
-  - watch
+  - '*'
```
Comment (kevincai) [inline deploy/operator.yaml:28]: is it a good practice to use "*" to acquire all permissions. one of the rule is acquiring only necessary permissions.

File: helm-charts/charts/kube-starrocks/charts/starrocks/templates/starrockscluster.yaml [hunk 6]
```diff
@@ -370,3 +370,33 @@ spec:
       configMapName: {{template  "starrockscluster.cn.configmap.name" . }}
       resolveKey: cn.conf
 {{- end }}
+{{- if .Values.starrocksFeProxySpec.enabled }}
+  starRocksFeProxySpec:
+    {{- if .Values.starrocksFeProxySpec.image.repository }}
```
Comment (kevincai) [inline helm-charts/charts/kube-starrocks/charts/starrocks/templates/starrockscluster.yaml:387]: what if use doesn't provide the repository setting, the crd will not have the image configuration, will operator fill the `image` field?

File: deploy/operator.yaml (hunk 5 above)
Comment (yandongxiao) [reply deploy/operator.yaml:28]: yes, you are right, but the original permissions are equal '*'.

File: helm-charts/charts/kube-starrocks/charts/starrocks/templates/starrockscluster.yaml (hunk 6 above)
Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/charts/starrocks/templates/starrockscluster.yaml:387]: in starrocksFeProxySpec, no fields is necessary.

File: helm-charts/charts/kube-starrocks/values.yaml (hunk 4 above)
Comment (yandongxiao) [reply helm-charts/charts/kube-starrocks/values.yaml:69]: Ok, I will create a new PR for testing for v1.8.0

File: pkg/sub_controller/feproxy/feproxy_configmap.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		return err
+	}
+	httpPort := rutils.GetPort(config, rutils.HTTP_PORT)
+
+	feServiceName := service.ExternalServiceName(src.Name, src.Spec.StarRocksFeSpec)
+	proxyPass := fmt.Sprintf("http://%s:%d", feServiceName, httpPort)
+
+	resolver := feProxySpec.Resolver
+	if resolver == "" {
+		resolver = "kube-dns.kube-system.svc.cluster.local"
```
Comment (kevincai) [inline pkg/sub_controller/feproxy/feproxy_configmap.go:36]: whether we can call api server to get the accurate core-dns service? e.g. the core-dns deployed in a different namespace, or the kubernetes has a different domain name other than `.cluster.local`

Comment (yandongxiao) [reply pkg/sub_controller/feproxy/feproxy_configmap.go:36]: I do not think api-server knows that, and that is why I add a resolver filed in CRD

**Review by kevincai — APPROVED** [review]
